### PR TITLE
XmlLoggingConfiguration: Set config values on correct LogFactory object

### PR DESCRIPTION
**Pull request by bhaeussermann**
Related to #1203

### CODE REVIEW

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 1]
```diff
@@ -472,12 +476,18 @@ private void ParseNLogElement(NLogXmlElement nlogElement, string filePath, bool
             if (filePath != null)
                 this.fileMustAutoReloadLookup[GetFileLookupKey(filePath)] = autoReload;
 
-            LogManager.ThrowExceptions = nlogElement.GetOptionalBooleanAttribute("throwExceptions", LogManager.ThrowExceptions);
+            if (customLogFactory == null)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: maybe we can pass LogManager as custom factory?

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 2]
```diff
@@ -472,12 +476,18 @@ private void ParseNLogElement(NLogXmlElement nlogElement, string filePath, bool
...
+                LogManager.ThrowExceptions = nlogElement.GetOptionalBooleanAttribute("throwExceptions", LogManager.ThrowExceptions);
+            else
+                customLogFactory.ThrowExceptions = nlogElement.GetOptionalBooleanAttribute("throwExceptions", customLogFactory.ThrowExceptions);
             InternalLogger.LogToConsole = nlogElement.GetOptionalBooleanAttribute("internalLogToConsole", InternalLogger.LogToConsole);
             InternalLogger.LogToConsoleError = nlogElement.GetOptionalBooleanAttribute("internalLogToConsoleError", InternalLogger.LogToConsoleError);
             InternalLogger.LogFile = nlogElement.GetOptionalAttribute("internalLogFile", InternalLogger.LogFile);
             InternalLogger.LogLevel = LogLevel.FromString(nlogElement.GetOptionalAttribute("internalLogLevel", InternalLogger.LogLevel.Name));
-            LogManager.GlobalThreshold = LogLevel.FromString(nlogElement.GetOptionalAttribute("globalThreshold", LogManager.GlobalThreshold.Name));
+            if (customLogFactory == null)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: Idem

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 3]
```diff
@@ -82,21 +84,23 @@ private ConfigurationItemFactory ConfigurationItemFactory
...
-        public XmlLoggingConfiguration(string fileName)
-        {
-            using (XmlReader reader = CreateFileReader(fileName))
-            {
-                this.Initialize(reader, fileName, false);
-            }
-        }
+        /// <param name="customLogFactory">The custom <see cref="LogFactory"/> to which to apply any applicable configuration values,
+        /// or <c>null</c> to use the global <see cref="LogFactory"/> (via <see cref="LogManager"/>).</param>
+        public XmlLoggingConfiguration(string fileName, LogFactory customLogFactory = null)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: Looks good but optional parameters aren't binary compatible. So we need extra ctors and should not alter the current ones.

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 4]
```diff
@@ -82,21 +84,23 @@ private ConfigurationItemFactory ConfigurationItemFactory
...
 
         /// <summary>
         /// Initializes a new instance of the <see cref="XmlLoggingConfiguration" /> class.
         /// </summary>
         /// <param name="fileName">Configuration file to be read.</param>
         /// <param name="ignoreErrors">Ignore any errors during configuration.</param>
-        public XmlLoggingConfiguration(string fileName, bool ignoreErrors)
+        /// <param name="customLogFactory">The custom <see cref="LogFactory"/> to which to apply any applicable configuration values,
+        /// or <c>null</c> to use the global <see cref="LogFactory"/> (via <see cref="LogManager"/>).</param>
+        public XmlLoggingConfiguration(string fileName, bool ignoreErrors, LogFactory customLogFactory = null)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: idem binary compatiblity

File: tests/NLog.UnitTests/LogFactoryTests.cs [hunk 5]
```diff
@@ -112,6 +109,38 @@ public void InvalidXMLConfiguration_ThrowErrorWhen_ThrowExceptionFlagIsSet()
         }
 
         [Fact]
+        public void SecondaryLogFactoryDoesNotTakePrimaryLogFactoryLock()
```
Comment (304NotModified) [inline tests/NLog.UnitTests/LogFactoryTests.cs]: What are where asserting here? no deadlock?

**Comment by 304NotModified** [comment]
Looks good. Added some notes. Binary compat  is the most important one.

File: tests/NLog.UnitTests/LogFactoryTests.cs (hunk 5 above)
Comment (bhaeussermann) [inline tests/NLog.UnitTests/LogFactoryTests.cs]: Yes. We run the thread while holding the lock of the global `LogFactory`. So if the code in the thread takes the same lock, we will deadlock and the `Join()` method will time out.

Comment (bhaeussermann) [inline tests/NLog.UnitTests/LogFactoryTests.cs]: Hence the test simulates what would in some cases cause a deadlock.

Comment (304NotModified) [inline tests/NLog.UnitTests/LogFactoryTests.cs]: Ow yes indeed. Please add that as comments to the method / lines. :)

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (bhaeussermann) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: Good idea!

**Comment by bhaeussermann** [comment]
AppVeyor:
```
LogManager.cs(137,24): error CS1574: Warning as Error: XML comment on 'NLog.LogManager.Configuration' has cref attribute 'LogFactory.Configuration' that could not be resolved [C:\projects\nlog\src\NLog\NLog.Xamarin.Android.csproj]
```

Wha?! I didn't touch this `cref` attribute and the `LogFactory.Configuration` property was not moved or renamed. Any ideas?

**Comment by 304NotModified** [comment]
> Wha?! I didn't touch this cref attribute and the LogFactory.Configuration property was not moved or renamed. Any ideas?

A `using` has been removed?

**Comment by bhaeussermann** [comment]
@304NotModified I cannot see what is wrong. If I right-click on any of these `cref`s in Visual Studio and select "Go To Definition", it takes me to the corresponding member. So they are still correct according to Visual Studio.

**Comment by 304NotModified** [comment]
well I would recommend to merge with master locally and then check it.

**Comment by bhaeussermann** [comment]
> well I would recommend to merge with master locally and then check it.

The `cref`s still seem fine according to Visual Studio. I cannot see what changes in this PR might cause the `cref`s to break.

**Comment by 304NotModified** [comment]
Will check it locally tomorrow. Strange issue. Have to be a missing `using` or something like that.

**Comment by 304NotModified** [comment]
https://github.com/bhaeussermann/NLog/pull/1

some ambiguity about Conguration (class name vs property)

**Comment by 304NotModified** [comment]
Thanks!

**Pull request merged**